### PR TITLE
Misc. updates for NED

### PR DESCRIPTION
**Pull request by Mehrad0711**

### CODE REVIEW

File: genienlp/arguments.py
Comment (s-jse) [inline genienlp/arguments.py]: `--ned_type_normalization` with `choices=['off', 'lenient', 'strict']` makes it more obvious what this argument does.

File: genienlp/ned/ned_utils.py
Comment (s-jse) [inline genienlp/ned/ned_utils.py]: Probably cleaner if this list of banned words is kept in a separate json file.

File: genienlp/arguments.py
Comment (gcampax) [inline genienlp/arguments.py]: I am like 50% confident that I saw this argument in Bootleg code as well, so why this help text talks about ngram-based lookup?

File: genienlp/arguments.py
Comment (gcampax) [inline genienlp/arguments.py:441]: What's the equivalent of the previous option here? The config backcompat code says "yes", but I thought the previous logic was "force"

File: genienlp/ned/bootleg.py
Comment (gcampax) [inline genienlp/ned/bootleg.py]: Uh? Why is this here? In the code?

File: genienlp/ned/bootleg.py
Comment (gcampax) [inline genienlp/ned/bootleg.py]:
```
if is_banned(alias):
    return
```

and you save one indentation level everywhere, making things more readable.

File: genienlp/ned/scripts/analyze_bootleg_results.py
Comment (gcampax) [inline genienlp/ned/scripts/analyze_bootleg_results.py]: This should be a command-line argument. You cannot assume you can write to the current directory.

File: genienlp/ned/scripts/analyze_bootleg_results.py
Comment (gcampax) [inline genienlp/ned/scripts/analyze_bootleg_results.py]: Why is this not a normal genienlp command?

File: genienlp/run_bootleg.py
Comment (gcampax) [inline genienlp/run_bootleg.py]: "ngram-based lookup" is implementation detail, the point is that entities shorter than that will not be retrieved

File: genienlp/ned/bootleg.py
Comment (gcampax) [inline genienlp/ned/bootleg.py]: This is not the place for this stuff.

Also, to have this stuff at all in Bootleg is meaningless: you're just raising your accuracy on the dev set by looking at oracle results and cleaning up by hand, it's not scientific.

File: tests/test_NED.sh
Comment (gcampax) [inline tests/test_NED.sh]: Mistake here?

**Review by gcampax — CHANGES_REQUESTED** [review]
Btw, as a general rule, you should rebase/squash patches like "make X work again" or "fixed to make tests pass" into whatever patch broke things. Ideally, tests should pass after _every_ commit, so if there are regressions not caught by the tests we can bisect and test all commits to find the culprit.

File: genienlp/ned/scripts/oracle_vs_bootleg.py
Comment (s-jse) [inline genienlp/ned/scripts/oracle_vs_bootleg.py]: Should this and `analyze_bootleg_results` be added as genienlp scripts?

**Review by s-jse — APPROVED** [review]

File: genienlp/arguments.py
Comment (Mehrad0711) [reply genienlp/arguments.py]: Yes. This argument is shared between the two scripts. Here it's used for the naive NED approach which does NER right before training (as opposed to bootleg which does it in a separate step by running run_bootleg.py)

File: genienlp/arguments.py
Comment (Mehrad0711) [reply genienlp/arguments.py:441]: previous approach was using "yes" logic.

File: genienlp/ned/bootleg.py
Comment (Mehrad0711) [reply genienlp/ned/bootleg.py]: `qid2typenames` map qids to an unordered list of types. I noticed the type we want is sometimes not ranked highly. So I have to do some reordering. I don't want to modify the files directly cause I'd forget what I changed later. Instead do it in the code for visibility and version tracking.

File: genienlp/ned/scripts/analyze_bootleg_results.py
Comment (Mehrad0711) [reply genienlp/ned/scripts/analyze_bootleg_results.py]: These scripts are for local data analysis. I don't want to crowd genienlp commands with data processing/ analysis scripts.

File: genienlp/ned/bootleg.py
Comment (Mehrad0711) [reply genienlp/ned/bootleg.py]: I'm reaching the conclusion that for certain entities we need to map them directly to the correct type.
It happens in 2 cases:
1) bootleg partly detected the entity and thus got the Qid wrong. 
2) bootleg alias2qids has incorrect entries where the book’s writer Qid is associated with the book name instead of the book itself, thus mapping to “person” type instead of “book”.

In none of those cases changing type mapping will fix the problem. Also, this mapping is only done for training set, not dev/ test set.
I added this as if-else in the code to see how experiments turn out. I'm eventually gonna move it to a json file similar to type mapping.

File: genienlp/ned/scripts/oracle_vs_bootleg.py
Comment (Mehrad0711) [reply genienlp/ned/scripts/oracle_vs_bootleg.py]: These scripts are for local data analysis. I don't want to crowd genienlp commands with data processing/ analysis scripts.

**Comment by Mehrad0711** [comment]
> Btw, as a general rule, you should rebase/squash patches like "make X work again" or "fixed to make tests pass" into whatever patch broke things. Ideally, tests should pass after _every_ commit, so if there are regressions not caught by the tests we can bisect and test all commits to find the culprit.

Believe me, I tried hard to make these PRs presentable and keep a clean commit history! In some cases, there were multiple fixes in one commit and it was hard (not impossible) to break it and merge with commits that introduced a bug. 
I agree we should strive for buildable commits though esp. now that more people will be contributing to the code.

File: genienlp/arguments.py
Comment (gcampax) [reply genienlp/arguments.py]: This is confusing. How about "Minimum token-length of entities (does not apply to Bootleg)"?

File: genienlp/ned/bootleg.py
Comment (gcampax) [reply genienlp/ned/bootleg.py]: This is the wrong way around it. Changing the code is harder. We don't want hacks in the code that confuse readers and contributors. If you want to version the bootleg input files, you can do it in the workdirs (with some light-weight override file, not by tracking the full bootleg file)

File: genienlp/ned/scripts/analyze_bootleg_results.py
Comment (gcampax) [reply genienlp/ned/scripts/analyze_bootleg_results.py]: This is a generally usable script that could be part of everyone's workflow. It's not a custom one-off script. I think it's worth making it part of genienlp proper.

File: genienlp/ned/bootleg.py
Comment (gcampax) [reply genienlp/ned/bootleg.py]: Yeah don't merge this hacks to the code. Add an override file if you need to, but I still don't think you need to.

File: genienlp/ned/scripts/analyze_bootleg_results.py
Comment (Mehrad0711) [reply genienlp/ned/scripts/analyze_bootleg_results.py]: My main concern is this will be an invitation for more scripts. As a general rule, I think we should exclude data manipulation scripts that do not rely on a specific model/ tokenizer and have them in genie-toolkit or workdirs. I can add these for now since you and Sina are both in favor and Kevin may wanna take a look too.

Comment (gcampax) [reply genienlp/ned/scripts/analyze_bootleg_results.py]: I agree, anything that's project/model specific should go into genie-workdirs.

**Review by gcampax — APPROVED** [review]
Ok to merge when tests are happy

**Pull request merged**
